### PR TITLE
Web Cache Deception Vulnerability on Go Frameworks

### PR DESCRIPTION
**Pull request by aydinnyunus**
#### Overview
This pull request introduces a new CodeQL query to detect potential Web Cache Deception vulnerabilities in web applications. Web Cache Deception is a security vulnerability where attackers trick a server into caching sensitive information, which they can later access. This query aims to identify code patterns that might make an application susceptible to this type of attack.

#### Changes Introduced
- **New Query Added**: `WebCacheDeceptionFiber.ql` - Detects patterns where web applications might cache sensitive information inadvertently.
- **New Query Added**: `WebCacheDeceptionGoChi.ql` - Detects patterns where web applications might cache sensitive information inadvertently.
- **Experimental Code Samples**: Included in the `/examples` directory, demonstrating both vulnerable (bad) and secure (good) coding practices related to caching.
- **Documentation**: Added documentation explaining the query's purpose, usage, and the nature of the vulnerability.

#### Implementation Details
- The query looks for server configurations or code patterns where URLs can be manipulated to cache sensitive information.
- Focuses on common web languages and frameworks where this vulnerability might occur.

#### Testing and Validation
- Query tested against a range of synthetic code samples (included in the PR).
- Validated for false positive rates and performance impact on large codebases.

#### Future Work
- Plan to extend support to additional languages and frameworks.
- Open to community feedback for further refinement and optimization.

#### References
- Link to the OWASP page on Web Cache Deception: [OWASP Web Cache Deception Attack](https://owasp.org/www-community/attacks/Web_Cache_Deception).

### CODE REVIEW

**Comment by aydinnyunus** [comment]
Probably I should not push the "vendor" directory. If you approve it I can remove it.

**Comment by owen-mc** [comment]
Please use [depstubber](https://github.com/github/depstubber) to create just the stubs that are needed and put those in the vendor directory.

**Comment by aydinnyunus** [comment]
Updated @owen-mc

File: go/ql/src/experimental/CWE-525/WebCacheDeceptionFiber.ql [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+ * @tags security
+ *       external/cwe/cwe-525
+ */
+
+import go
+
+from DataFlow::CallNode httpHandleFuncCall, ImportSpec importSpec
+where
+  importSpec.getPath() = "github.com/gofiber/fiber/v2" and
+  httpHandleFuncCall.getCall().getArgument(0).toString().matches("%/*%") and
```
[the remark this answers is not in the document]
Comment (aydinnyunus) [reply go/ql/src/experimental/CWE-525/WebCacheDeceptionFiber.ql:18]: `getStringValue` does not works for me, so I use `toString`

File: go/ql/src/experimental/CWE-525/WebCacheDeceptionFiber.ql [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+  not httpHandleFuncCall.getCall().getArgument(0).toString().matches("%$%") and
```
[the remark this answers is not in the document]
Comment (aydinnyunus) [reply go/ql/src/experimental/CWE-525/WebCacheDeceptionFiber.ql:19]: `getStringValue` does not works for me, so I use `toString`

**Comment by aydinnyunus** [comment]
fixed the qhelp issue and tests.

File: go/ql/src/experimental/CWE-525/WebCacheDeceptionGoChi.ql [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+ * @tags security
+ *       external/cwe/cwe-525
+ */
+
+import go
+
+from DataFlow::CallNode httpHandleFuncCall, ImportSpec importSpec
+where
+  importSpec.getPath() = "github.com/go-chi/chi/v5" and
+  httpHandleFuncCall.getCall().getArgument(0).toString().matches("%/*%") and
```
[the remark this answers is not in the document]
Comment (aydinnyunus) [reply go/ql/src/experimental/CWE-525/WebCacheDeceptionGoChi.ql]: `getStringValue` does not works for me, so I use `toString`

File: go/ql/src/experimental/CWE-525/WebCacheDeceptionGoChi.ql [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+  not httpHandleFuncCall.getCall().getArgument(0).toString().matches("%$%") and
```
[the remark this answers is not in the document]
Comment (aydinnyunus) [reply go/ql/src/experimental/CWE-525/WebCacheDeceptionGoChi.ql]: `getStringValue` does not works for me, so I use `toString`

File: go/ql/src/experimental/CWE-525/WebCacheDeceptionFiber.ql (hunk 2 above)
[the remark this answers is not in the document]
Comment (owen-mc) [reply go/ql/src/experimental/CWE-525/WebCacheDeceptionFiber.ql:19]: What string does `getStringValue` give you? Or does it not have a value in a case that you care about? Queries cannot use `toString`.

[the remark this answers is not in the document]
Comment (aydinnyunus) [reply go/ql/src/experimental/CWE-525/WebCacheDeceptionFiber.ql:19]: I don't know why but I did not get the endpoint which is` "/*"`. I use `type()` for getting type but it is not string for CodeQL.

**Code Sample:** 

```
r.Get("/*", func(w http.ResponseWriter, r *http.Request) {
		w.Write([]byte("welcome"))
	})
```

When I use `getStringValue` it did not get this argument.

**Comment by owen-mc** [comment]
We already have a query for web cache deception (which you contributed). You shouldn't introduce a new query for each framework - we would end up with too many queries. Instead, you should adapt the existing query to be able to cover all of the frameworks. A common way to do this is to use an abstract class and for each framework you have a class extending it containing the framework-specific logic. The query can then use the abstract class.

**Comment by aydinnyunus** [comment]
> We already have a query for web cache deception (which you contributed). You shouldn't introduce a new query for each framework - we would end up with too many queries. Instead, you should adapt the existing query to be able to cover all of the frameworks. A common way to do this is to use an abstract class and for each framework you have a class extending it containing the framework-specific logic. The query can then use the abstract class.

I don't think that one rule can cover all of the frameworks because it depends on the framework behaviour. I need to create different checks for different frameworks.

## For Example

My Previous Rule checks for endpoint ends with "%/" character. But in this frameworks I need to check "%*". 

I did not find a solution to do it. How can I do it on one rule ?

**Comment by owen-mc** [comment]
There are various methods in the CodeQL language to deal with this kind of problem. The one that is most applicable here is abstract classes. If you aren't familiar with them you can read [the documentation here](https://codeql.github.com/docs/ql-language-reference/types/#abstract-classes). It may also be helpful to look at examples of them being used. They are used in almost every query, but I have tried to find a simple case that is similar to this PR. [Here](https://github.com/github/codeql/blob/main/go/ql/src/experimental/CWE-321/HardcodedKeysLib.qll#L24) is the definition of an abstract class called `Sink`. It is used in two places, [here](https://github.com/github/codeql/blob/main/go/ql/src/experimental/CWE-321/HardcodedKeysLib.qll#L374) and [here](https://github.com/github/codeql/blob/main/go/ql/src/experimental/CWE-321/HardcodedKeysLib.qll#L382). The only other mention of it is in a number of (non-abstract) classes which extend `Sink`, and which contain the logic for a sink for that query from a particular library, like [GolangJwtSign](https://github.com/github/codeql/blob/main/go/ql/src/experimental/CWE-321/HardcodedKeysLib.qll#L53), [KatarasJwt](https://github.com/github/codeql/blob/main/go/ql/src/experimental/CWE-321/HardcodedKeysLib.qll#L81), [IrisJwt](https://github.com/github/codeql/blob/main/go/ql/src/experimental/CWE-321/HardcodedKeysLib.qll#L110), and many more. So those two uses of `Sink` refer to any element of any of the classes extending `Sink`.

**Comment by aydinnyunus** [comment]
Thank you so much @owen-mc. You are the best. Should I need to close this request or continue on this ?

**Comment by owen-mc** [comment]
I would keep this PR open. It is linked with the bug bounty application.

**Comment by aydinnyunus** [comment]
Okay. I create abstract classes but did not get any match. What am I missing ?

Also I tried with extends CallNode and basic queries without checking endpoints but none of them matched.

`abstract class Sink extends DataFlow::CallNode`

File: go/ql/src/experimental/CWE-525/WebCacheDeceptionLib.qll [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+    override predicate isSink(DataFlow::Node sink) { sink instanceof Sink }
+  }
+
+  private module Config implements DataFlow::ConfigSig {
+    predicate isSource(DataFlow::Node source) { source instanceof Source }
+
+    predicate isSink(DataFlow::Node sink) { sink instanceof Sink }
+  }
+
+  module Flow = TaintTracking::Global<Config>;
```
Comment (owen-mc) [inline go/ql/src/experimental/CWE-525/WebCacheDeceptionLib.qll:67]:
```suggestion
  abstract class Sink extends DataFlow::Node { }

  private class GoNetHTTP extends Sink {
    GoNetHTTP() {
      exists(
        DataFlow::CallNode m, DataFlow::ReadNode rn, Http::HeaderWrite::Range hw, DeclaredFunction f
      |
        m.getTarget().hasQualifiedName("net/http", "HandleFunc") and
        m.getArgument(0).getStringValue().matches("%/") and
        m.getArgument(1) = rn and
        rn.reads(f) and
        f.getParameter(0) = hw.getResponseWriter() and
        hw.getHeaderName() = "cache-control" and
        this = m.getArgument(0)
      )
    }
  }

  private class GoFiber extends Sink {
    GoFiber() {
      exists(ImportSpec i |
        i.getPath() = "github.com/gofiber/fiber" or
        i.getPath() = "github.com/gofiber/fiber/v2"
      |
        exists(DataFlow::MethodCallNode m |
          m.getCall().getArgument(0).toString().matches("%/*%") and
          this = m.getArgument(0)
        )
      )
    }
  }

  private class GoChi extends Sink {
    GoChi() {
      exists(ImportSpec i |
        i.getPath() = "github.com/go-chi/chi/v5" or
        i.getPath() = "github.com/go-chi/chi/v5/middleware"
      |
        exists(DataFlow::MethodCallNode m |
          m.getCall().getArgument(0).toString().matches("%/*%") and
          this = m.getArgument(0)
        )
      )
    }
  }
```

File: go/ql/src/experimental/CWE-525/WebCacheDeception.ql [hunk 6]
```diff
@@ -1,6 +1,6 @@
 /*
  * @name Web Cache Deception
- * @description A caching system has been detected on the application and is vulnerable to web cache deception. By manipulating the URL it is possible to force the application to cache pages that are only accessible by an authenticated user. Once cached, these pages can be accessed by an unauthenticated user.
+ * @description A caching system has been detected on the application and is vulnerable to web cache deception on Gofiber. By manipulating the URL it is possible to force the application to cache pages that are only accessible by an authenticated user. Once cached, these pages can be accessed by an unauthenticated user.
```
Comment (owen-mc) [inline go/ql/src/experimental/CWE-525/WebCacheDeception.ql:3]:
```suggestion
 * @description A caching system has been detected on the application and is vulnerable to web cache deception. By manipulating the URL it is possible to force the application to cache pages that are only accessible by an authenticated user. Once cached, these pages can be accessed by an unauthenticated user.
```

File: go/ql/src/experimental/CWE-525/WebCacheDeception.ql [hunk 7]
```diff
@@ -11,17 +11,10 @@
  */
 
 import go
+import WebCacheDeceptionLib
+import WebCacheDeception::Flow::PathGraph
```
Comment (owen-mc) [inline go/ql/src/experimental/CWE-525/WebCacheDeception.ql:15]:
```suggestion
```

File: go/ql/src/experimental/CWE-525/WebCacheDeception.ql [hunk 8]
```diff
@@ -11,17 +11,10 @@
...
-  rn.reads(f) and
-  f.getParameter(0) = hw.getResponseWriter() and
-  hw.getHeaderName() = "cache-control"
-select httpHandleFuncCall.getArgument(0),
-  "Wildcard Endpoint used with " + httpHandleFuncCall.getArgument(0) + " and '" + hw.getHeaderName()
-    + "' Header is used"
+from WebCacheDeception::Flow::PathNode source, WebCacheDeception::Flow::PathNode sink
+where WebCacheDeception::Flow::flowPath(source, sink)
+select sink.getNode(), source, sink, "$@ is used as wildcard endpoint.", source.getNode(),
+  "Web Cache Deception"
```
Comment (owen-mc) [inline go/ql/src/experimental/CWE-525/WebCacheDeception.ql:20]:
```suggestion
from WebCacheDeception::Sink httpHandleFuncCall
select httpHandleFuncCall, "$@ is used as wildcard endpoint.", httpHandleFuncCall.getNode(),
  "Web Cache Deception"
```

**Review by owen-mc — CHANGES_REQUESTED** [review]
You don't need all the data-flow stuff, because you aren't looking for a flow path in this query.

**Comment by owen-mc** [comment]
One of your tests is failing.

**Comment by aydinnyunus** [comment]
I think it is okay now. Can you restart the tests ?

**Comment by aydinnyunus** [comment]
I think it is passed the pipeline but the toString problems still exists.

**Comment by owen-mc** [comment]
Every time you merge main in I can no longer see the results of the tests. Please do not do it so often, unless it is needed because of a merge conflict.

**Comment by aydinnyunus** [comment]
Okay 🙏

**Review by owen-mc — APPROVED** [review]

**Pull request merged**